src/Cryptol/Parser/AST.hs

```diff
@@ -495,6 +496,45 @@ data Bind name = Bind
   , bExport    :: !ExportType
   } deriving (Eq, Generic, NFData, Functor, Show)
 
+bindParams :: Bind name -> [Pattern name]
+bindParams b = case bParams b of
+  PatternParams ps -> ps
+  DroppedParams _ _ -> []
+
+-- | Sets the number of parameters for a binding to zero, noting
+-- the original number and source location of the patterns.
+-- e.g. when rewriting @let f a b c = \x -> ...@, into
+-- @let f = \\a b c x -> ...@ the parameters for @f@ change from
+-- @PatternParams [a,b,c]@ to @DroppedParams (getLoc [a,b,c]) 3@
+dropParams :: BindParams name -> BindParams name
+dropParams bps = case bps of
+  PatternParams ps -> DroppedParams (getLoc ps) (length ps)
+  DroppedParams rng i -> DroppedParams rng i
+
+-- | Range encompassing the LHS of a binder, its signature, but not
+-- its definition.
+bindHeaderLoc :: Bind name -> Maybe Range
+bindHeaderLoc b = getLoc (bName b, (bSignature b, bParams b))
+
+-- | An empty 'BindParams' (i.e. zero parameters).
+--   Note that 'dropParams' should be used instead of this
+--   when rewriting an existing 'Bind' to have no parameters.
+noParams :: BindParams name
+noParams = PatternParams []
+
+data BindParams name = 
```

Comment:
If I understand, this data type really only exists for the sake of improving error messages. Is that right? If so, it would be worth stating that in a comment.